### PR TITLE
Allow downloading of desktop embedding artifacts

### LINKED ISSUE #30663: Update flutter cache/precache to support desktop artifacts
This should include separate flags for macos, windows, and linux and download the current artifact shells similarly to the other artifact types. Now that we can handle lazy artifacts, this does not regress download times for other users.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Adds desktop embedding shells to cache logic, and allows downloading via the precache command.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/30663

## Tests

I added the following tests:

- precache with all flags requests all artifacts off stable.
- precache with all flags requests only stable platforms on stable branch

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by jonahwilliams** [comment]
This should not land until https://github.com/flutter/flutter/pull/30153 to prevent this from leaking into stable.

File: packages/flutter_tools/lib/src/cache.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/cache.dart]: How about something like:
```
final List<String> frameworkNames = ['Flutter', 'FlutterMacOS'];
for (frameworkName in frameworkNames) {
  final File frameworkZip = fs.file(fs.path.join(dir.path, '$frameworkName.framework.zip'));
  ...
}
```
rather than duplicating the code.

File: packages/flutter_tools/lib/src/cache.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/cache.dart]: Since I'm not familiar with the overall structure of this code: why does the macOS framework need to be specified/handled in two different places? It's both here and in the artifact list later. The iOS code doesn't seem to have the same structure (it's not obvious to me why Flutter.framework has special handling here, but it's not in the list below like the macOS version is.)

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/cache.dart]: Its a bit of a historical artifact (har-har), I can't quite remove it yet but eventually this will be gone. TLDR: I am scared to touch it

File: packages/flutter_tools/lib/src/cache.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/cache.dart]: Done

File: packages/flutter_tools/lib/src/commands/precache.dart
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/precache.dart]: This check is reversed.

**Review by stuartmorgan-g — COMMENTED** [review]
I cloned this to try it out (with the fix for the check below); `--macos` is working for me, but `--linux` and `--windows` appear to be no-ops; nothing prints out, and I can't find any `.dll` or `.so` files in `bin/cache`, nor the client wrapper source. Is it filtering by my host platform?

File: packages/flutter_tools/lib/src/commands/precache.dart
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/precache.dart]: Perhaps I should add a test...

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/precache.dart]: Test added.

**Comment by jonahwilliams** [comment]
Right now this is filtering by host platform, so macos artifacts are only downloaded for mac. If that isn't correct, I can remove that restriction.

File: packages/flutter_tools/test/commands/precache_test.dart
Comment (stuartmorgan) [inline packages/flutter_tools/test/commands/precache_test.dart]: Typo: Release

File: packages/flutter_tools/test/commands/precache_test.dart
Comment (stuartmorgan) [inline packages/flutter_tools/test/commands/precache_test.dart]: The indentation for everything starting at this line seems off.

File: packages/flutter_tools/test/commands/precache_test.dart
Comment (stuartmorgan) [inline packages/flutter_tools/test/commands/precache_test.dart]: It took me a couple of reads to realize that this mock was only applied in the second test's overrides; I was confused at first to how the test was passing. Maybe move the declaration and setup of `flutterVersion` down to between the two tests to make it clearer where it's used?

**Review by stuartmorgan-g — APPROVED** [review]
LGTM with nits. (Caveat: I'm not familiar with the Dart testing infrastructure yet, so my review of the tests is probably useful. Up to you if you want a second reviewer for that part.)

**Pull request merged**
